packages/eslint-plugin/tests/rules/no-base-to-string.test.ts

```diff
@@ -535,6 +660,51 @@ declare const foo: Bar & Foo;
         },
       ],
     },
+    {
+      code: `
+        class A {}
+        String([new A(), 'str']);
```

Comment:
Interesting, I didn't know an empty class counts as the dreaded `{}`.

I've relied on some existing tests from the recently added `join()` case (added in #10287) as it's somewhat similar to the issue this PR tackles. From what I can see, the tests for `join()` also have the same problem ([example here](https://github.com/typescript-eslint/typescript-eslint/blob/7defffef8fe77387a22dbd5a538ff4cafcf81253/packages/eslint-plugin/tests/rules/no-base-to-string.test.ts#L523-L537)).

Would it make sense to look into or adjust the existing tests too?

Reply:
> Interesting! I didn't know an empty class counts as the dreaded `{}`.

There are so many bizarre JS/TS trivia you get to learn when working on the linter 😆

> I've relied on some existing tests from the recently added `join()` case (added in #10287) as it's somewhat similar to the issue this PR tackles. From what I can see, some of the tests for `join()` also have the same problem ([one example here](https://github.com/typescript-eslint/typescript-eslint/blob/7defffef8fe77387a22dbd5a538ff4cafcf81253/packages/eslint-plugin/tests/rules/no-base-to-string.test.ts#L523-L537)).
> 
> Would it make sense to look into and potentially adjust the existing tests too?

my opinion - now that we've seen I'd consider it required for added test cases and optional extra credit for existing ones, depending on your level of motivation 😁.

Reply:
I have a question I'm not sure about. I'll love to hear your opinion about it: To my understanding, the test cases for `array.join()`, `array.toString`, `String(array)`, and `${[{}, {}]};` are very similar or even identical. I'm considering using a helper function like some rules do for similar issues (examples: [1](https://github.com/typescript-eslint/typescript-eslint/blob/d24a82854d06089cbd2a8801f2982fd4781f3701/packages/eslint-plugin/tests/rules/sort-type-constituents.test.ts#L91-L320), [2](https://github.com/typescript-eslint/typescript-eslint/blob/d24a82854d06089cbd2a8801f2982fd4781f3701/packages/eslint-plugin/tests/rules/no-unsafe-assignment.test.ts#L17-L61)), so that all cases are covered for each way of stringifying an array.

What's your opinion on this? I think it's convenient and helps make the test more concise, but it does make the cases somewhat harder to understand.

Reply:
Hmm. I'm pretty anti- test case helpers... I think @JoshuaKGoldberg tends to be as well. Though I understand why you'd want to use one for a case like this. But I'm still negative on it 😬 Makes writing easier but reviewing harder

Reply:
Yeah I'm pretty -1 on test case helpers. They're much less analyzable than direct raw strings.

IME most of the time autogenerated tests either:
* Only automate up to about a half dozen things or so: which is not worth the automation on its own
* Automate including very many combinations of cases together: which is an anti-pattern of overloading too many duplicate cases across tests

Reply:
Hmm I tend to agree! I've updated the PR to match this.

Reply:
Thanks for dealing with this! 🙏  It really did make it so much easier to review the test cases ❤️